### PR TITLE
For discussion: full-file hashes in hyperdrive metadata

### PR DESCRIPTION
**Pull request by bnewbold**
[Rendered pre-merge](https://github.com/bnewbold/dat-deps/blob/dep-hyperdrive-hashes/proposals/0000-hyperdrive-hashes.md)

"Full-file hashes are optionally included in hyperdrive metadata to complement the existing cryptographic-strength hashing of sub-file chunks. Multiple popular hash algorithms can be included at the same time."

Previous discussion:
- https://github.com/mafintosh/hyperdrive/issues/203
- https://github.com/datproject/discussions/issues/77#issuecomment-354969691

Before seriously reviewing to merge as Draft, I would want to demonstrate working code and have a better idea what the API would look like, but this is otherwise pretty flushed out.

cc: @martinheidegger

### CODE REVIEW

**Comment by pfrazee** [comment]
Other advantages:
 - Hashes can be used as pointers on the network which imply happens-before causality. That is: If a record can point to the hash of the file, and the record has not been changed since it was created, then we know that the pointed-to file existed before the record did.
 - The discovery & wire network can be expanded in the future to fetch individual files using these hashes.

File: proposals/0000-hyperdrive-hashes.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+What does the user-facing API look like, specifically?
+
+Should we allow non-standard hashes, like the git "hash", or higher-level
+references like (single-file) bittorrent magnet links or IPFS file references?
+
+Modifying a small part of a large file would require re-hashing the entire
+file, which is slow. Should we skip including the updated hashes in this case?
+Currently mitigated by the fact that we duplicate the entire file when recoding
+changes or additions.
```
Comment (martinheidegger) [inline proposals/0000-hyperdrive-hashes.md]: Hashes are optional so it is up to the implementor/user to add a hash or not. Including for updates of a file.

File: proposals/0000-hyperdrive-hashes.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  optional uint64 mtime = 8;
+  optional uint64 ctime = 9;
+  repeated ExtraHash hashes = 10;
+}
+```
+
+`type` is a number representing the hash algorithm, and `value` is the
+bytestring of the hash output itself. The length of the hash digest (in bytes)
+is available from protobuf metadata for the value. This scheme, and the `type`
+value table, is intended to be interoperable with the [multihash][multihash]
```
Comment (martinheidegger) [inline proposals/0000-hyperdrive-hashes.md]: We should reference the [multihash table](https://github.com/multiformats/multihash/blob/master/hashtable.csv) to specify what the types are?!

Comment (bnewbold) [reply proposals/0000-hyperdrive-hashes.md]: The table is linked from the multihash homepage linked... i'm wary of deep linking directly into a github blob (the repo could move to a new platform or file could be renamed), but maybe that's an overblown concern.

File: proposals/0000-hyperdrive-hashes.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+modified file.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+What does the user-facing API look like, specifically?
+
+Should we allow non-standard hashes, like the git "hash", or higher-level
+references like (single-file) bittorrent magnet links or IPFS file references?
```
[the remark this answers is not in the document]
Comment (bnewbold) [reply proposals/0000-hyperdrive-hashes.md]: I'm not sure I follow. What I was getting at here was "what if there are additional hash or merkel tree references a user would want to include that are not in the multihash table"?

File: proposals/0000-hyperdrive-hashes.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    md5         0x00D5
+    sha1        0x0011
+    sha2-256    0x0012
+    sha2-512    0x0013
+    blake2b-256 0xB220
+```
+
+Multiple hashes would be calculated in parallel with the existing
+chunking/hashing process, in a streaming fashion. Final hashes would be
+calculated when the chunking is complete, and included in the `Stat` metadata.
```
[the remark this answers is not in the document]
Comment (bnewbold) [reply proposals/0000-hyperdrive-hashes.md]: Correct; this is making the point that even with multiple hashes, the file only needs to be scanned (read from disk) only once.

File: proposals/0000-hyperdrive-hashes.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+Authors: [Bryan Newbold](https://github.com/bnewbold)
+
+
+# Summary
+[summary]: #summary
+
+Full-file hashes are optionally included in hyperdrive metadata to complement
+the existing cryptographic-strength hashing of sub-file chunks. Multiple
+popular hash algorithms can be included at the same time.
```
[the remark this answers is not in the document]
Comment (bnewbold) [reply proposals/0000-hyperdrive-hashes.md]: Yes! If we are trying to be inter-operable with existing databases and large users, and in particular to bridge to older "legacy" systems which might not support newer ("better") hash functions. To be transparent, I work at the Internet Archive, and we have dozens of petabytes of files hashed and cataloged with the MD5 and SHA1 hashes (because they are popular, not because they are "strong" or the "best" in any sense). We'll probably re-hash with new algorithms some day, but would like to do so only rarely.

File: proposals/0000-hyperdrive-hashes.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+separately; the chunk boundaries usually do not correspond to individual files.
+git prepends the size of the file (in bytes) as a fixed header before hashing
+and storing the file as a "blob". This makes comparison or interoperability
+between these systems impossible without having either a universal cross-hash
+table (infeasible to build in the general sense) or without having the full
+file contents on-hand to compare or re-hash in all three formats.
+
+The design decisions to adopt hash variants are usually well-founded, motivated
+by security concerns (such as pre-image attacks), efficiency, and
+implementation concerns.
```
[the remark this answers is not in the document]
Comment (bnewbold) [reply proposals/0000-hyperdrive-hashes.md]: I'm not sure I understand. The dat implementation *does* use something other than popular full-file hash algorithms internally, for good reasons.

**Comment by bnewbold** [comment]
> I feel like there are various recommendations missing from the dat project

This was intentional. I don't think we should be over prescriptive; folks should be able to adapt this feature to their own needs (including ones we aren't even thinking of at this time).

For the sake of simplicity, if users or implementors don't want to be bothered choosing or coordinating algorithms to use by default, this draft does say:

> For 2018, recommended default full-file hash functions to include are SHA1 (for popularity and interoperability) and blake2b-256 (already used in other parts of the Dat protocol stack).

File: proposals/0000-hyperdrive-hashes.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (martinheidegger) [reply proposals/0000-hyperdrive-hashes.md]: This was a little nitpicking around my impression that the sentence would have the same meaning and impact without "popular"; this blew out of proportion.

File: proposals/0000-hyperdrive-hashes.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (martinheidegger) [reply proposals/0000-hyperdrive-hashes.md]: It uses chunk-based hashes, yes. But all data received through the dat protocol is signed. I am not sure how you would smuggle unsigned content into a dat.

File: proposals/0000-hyperdrive-hashes.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (martinheidegger) [reply proposals/0000-hyperdrive-hashes.md]: Let me rephrase: No: we should not allow non-standard hashes. Though "standard" in this context means the standard we set. I am okay with a "githash" being added for example.

File: proposals/0000-hyperdrive-hashes.md (hunk 2 above)
Comment (martinheidegger) [reply proposals/0000-hyperdrive-hashes.md]: Having a link seems better than not having it. To remove/reduce that concern you could link to a commit number or make a mirror of it.

**Comment by martinheidegger** [comment]
I seemed to have overread that section :blush:

> For 2018, recommended default full-file hash functions to include are SHA1 (for popularity and interoperability) and blake2b-256 (already used in other parts of the Dat protocol stack).

Now I am all good - though @mafintosh might have something to say about additionally having to compute SHA1

**Comment by bnewbold** [comment]
Maybe "widely used"
